### PR TITLE
fix: validate non models when using object literal

### PR DESCRIPTION
**Pull request by dpilch**
#### Description of changes

Validate non model fields when using object literal syntax.

Non models can be created with an object literal instead of a constructor. When the object literal is used the validation is skipped. This results in `DataStore.save` successfully saving locally, but the subsequent mutation fails to sync with the cloud. The failed sync is pushed to `errorHandler`, but is otherwise lost to the user.

The following are equivalent methods of creating a new Post.

With constructor:

```
new Post({
  name: 'My Post',
  metadata: new Metadata({
    author: 'Some author',
    tags: undefined,
  }),
})
```

Without constructor:

```
new Post({
  name: 'My Post',
  metadata: { // no constructor
    author: 'Some author',
    tags: undefined,
  },
})
```

In the second example there is no validation. This would allow the user to enter something like `author: 5` without an easily noticeable failure. If they were to use TypeScript they would get a build failure. When not using TypeScript there is no error.

#### Issue #, if available

Resolves [#10357](https://github.com/aws-amplify/amplify-js/issues/10357)

#### Description of how you validated changes

* Unit testing
* Tested in an app with the following schema

```
type Parent @model {
  id: ID!
  name: String
  requiredChildren: [Child]!
  requiredChild: Child!
  children: [Child]
  child: Child
}

type Child {
  id: ID!
  name: String!
  description: String
}
```

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/__tests__/DataStore.ts [hunk 1]
```diff
@@ -1516,295 +1516,596 @@ describe('DataStore tests', () => {
...
 		});
 
-		test('Instantiation validations', async () => {
-			expect(() => {
-				new Model({
-					field1: undefined,
-					dateCreated: new Date().toISOString(),
-				});
-			}).toThrowError('Field field1 is required');
+		describe.only('Instantiation validations', () => {
```
Comment (svidgen) [inline packages/datastore/__tests__/DataStore.ts:1519]: You've got a `.only` hanging out here!

File: packages/datastore/src/datastore/datastore.ts [hunk 2]
```diff
@@ -394,6 +397,56 @@ const validateModelFields =
 						`Field ${name} should be of type ${type}, validation failed. ${v}`
 					);
 				}
+			} else if (isNonModelFieldType(type)) {
```
Comment (svidgen) [inline packages/datastore/src/datastore/datastore.ts:400]: Not a blocker ⬇️ 

Would be good to think about whether some of these larger methods can be elegantly decomposed down the line.

**Review by svidgen — APPROVED** [review]

File: packages/datastore/src/datastore/datastore.ts [hunk 3]
```diff
@@ -394,6 +397,56 @@ const validateModelFields =
...
+					const modelValidator = validateModelFields(subNonModelDefinition);
+
+					if (isArray) {
+						let errorTypeText: string = type.nonModel;
+						if (!isRequired) {
+							errorTypeText = `${type.nonModel} | null | undefined`;
+						}
+						if (!Array.isArray(v)) {
+							throw new Error(
+								`Field ${name} should be of type [${errorTypeText}], ${typeof v} received. ${v}`
```
Comment (david-mcafee) [inline packages/datastore/src/datastore/datastore.ts:414]: Nit (and not a blocker): could place these error messages [here](https://github.com/aws-amplify/amplify-js/blob/main/packages/datastore/src/util.ts#L54-L62)

**Review by david-mcafee — APPROVED** [review]
LGTM! Just left one minor comment.

**Pull request merged**